### PR TITLE
http_caldav_sched.c: sanitize CN before using it in To: header

### LINKED ISSUE #4955: Incorrect "From" parsing on CalDAV invites
KOrganizer creates CalDAV events with `ORGANIZER;CN=Anthony Messina <Anthony Messina>:MAILTO:
 amessina@example.com` (with the name duplicated inside <> the CN).  While this is unfortunate, Cyrus improperly parses the organizer and sends out invitations From: `"Anthony Messina <AnthonyMessina@example.com>" <amessina@example.com>`, compressing the name and appending the domain, resulting in `<AnthonyMessina@example.com>`

This creates issues for other clients (recipients), which attempt to send replies to `<AnthonyMessina@example.com>` instead of `<amessina@example.com>`

While I don't know why Korganizer duplicates the ORGANIZER name, Cyrus should not convert the CN into an email address for the From: of the invite.

Note, Cyrus does not alter the ORGANIZER in the attached invite itself.

### PR DESCRIPTION
**Pull request by rsto**
Fixes #4955

### CODE REVIEW

**Comment by rsto** [comment]
@elliefm This is meant to fix #4955 but could you please help me out getting the Cassandane test draft in the above PR implemented?

- To make Cyrus actually send iMIP messages with sendmail, the test has to disable the `imipnotifier` imapd.conf option. This is set to `imip` in Cassandane. I tried resetting that option with a test magic to Perl `undef`, but this seems to not have made that setting get deleted from the config.
- To assert this test, we want to validate that the `To:` MIME header of the message that gets sent to Cassandane's SMTPServer is sane. But I struggled to figure out how get the data out to the test code:
  - XmtpServer does have option `store_msg` to store the contents of the SMTP DATA command and the `xmtp_tmp_dir` option allows to set this to the test basedir `tmp` directory. But even after removing the `UNLINK => 1` parameters that XmtpServer uses when creating that spool with `tempfile` I just could not made it store the file there.
  - Probably simple, the Cassandane SMTPServer implementation can override the `output_body` function of its XmptServer parent, but then how would we get the contents of that from the SMTPServer instance to the test code?
- When I ran the Cassandane test of the PR with manually enabling Cyrus to use sendmail as imipnotifier, I noticed that the test times out when it attempts to send more than 3 messages, I guess that's related to the test runner creating 3 SMTPServer instances and then waiting until one comes available before allowing any further send. 

Do you have any ideas how to resolve the above issues? Thanks.

File: imap/http_caldav_sched.c [hunk 1]
```diff
@@ -202,9 +202,36 @@ static void add_address(struct address_t **recipients, icalproperty *prop,
     new->addr = address;
     param = icalproperty_get_first_parameter(prop, ICAL_CN_PARAMETER);
     if (param) {
-        new->name = icalparameter_get_cn(param);
-        new->qpname = charset_encode_mimeheader(new->name, 0, 0);
+        // RFC 5545 defines CN to "specify the common name" but at least
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_caldav_sched.c:205]: Isn't it KOrganizer that writes an angle-bracket name in the first place, like:
```
ORGANIZER;CN=Foo <Foo>:mailto:foo@bar.com
```
At least that's how I understood your issue description.

But I'll just remove the name-calling in the comment, we'll need to sanitize the value in Cyrus regardless.

[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_caldav_sched.c:205]: Isn't it KOrganizer that puts these angle-brackets in the CN parameter value? At least that's how I understood your issue description, e.g.
```
ORGANIZER;CN=Foo <Bar>:mailto:foo@example.com
```

But the name-calling in the comment is not necessary, we'll remove it.

**Comment by rsto** [comment]
Updated to produce either safe names, q-encoded names and quoted names.

**Comment by elliefm** [comment]
@rsto I've pulled your branch and I'm tinkering around...

> imipnotifier

I have a fix for this on my local copy of your branch

> validate To: header

I'm currently tinkering around in the guts of XmtpServer and SMTPServer to try and find a good way to capture this.  I don't have anything conclusive yet.

> test times out with more than three messages

This is a known problem when running Cassandane in verbose mode, but should go away if you **don't** use the `-v` flag.  Which is pretty annoying while debugging!  I've pinged you on an FM-internal issue where I wrote up the details of this a while ago.

I'm about to wrap up my work day, but I'll continue with this on Monday.

Also, please let me know if it's okay for me to push my work straight to your branch... I believe this should be possible, if you haven't unchecked the "allow maintainers to yadda yadda" box on the PR, but I've never tried it before.

**Comment by elliefm** [comment]
@rsto I've got this mostly working I think, just waiting on a full cassandane run to finish to check if I broke anything along the way...

* fixed Instance to not force "imipnotifier: imip", and made the few tests/suites that actually want this behaviour ask for it themselves (including Caldav)
* made SMTPServer save any messages it receives in $basedir/smtpd so they can be inspected later
* made the imip_quote_cn test override the "imipnotifier: imip" it inherited from the Caldav suite before starting up (which is possible now that Instance doesn't force it)
* made the imip_quote_cn test validate the To headers by examining the message files created by SMTPServer

**Comment by rsto** [comment]
> I've got this mostly working I think, just waiting on a full cassandane run to finish to check if I broke anything along the way...

Great, looking forward to it!

**Comment by rsto** [comment]
Now updated with additional test input.

**Comment by elliefm** [comment]
@rsto I've rebased my work on your latest update, and made a draft PR against your fork with my changes: https://github.com/rsto/cyrus-imapd/pull/1

**Review by ksmurchison — APPROVED** [review]

**Pull request merged**